src/backend/metal/src/command.rs

```diff
@@ -159,14 +159,19 @@ struct State {
     resources_cs: StageResources,
     index_buffer: Option<IndexBuffer>,
     attribute_buffer_index: usize,
+    rasterizer_state: Option<native::RasterizerState>,
     depth_stencil_state: Option<metal::DepthStencilState>,
+    push_constant_data: HashMap<native::UniquePipelineLayoutId, Vec<u32>>,
```

Comment:
I'm a bit skeptical of this design here:
* How does it handle the case where a user uses the **same** layout for compute and graphics, but pushes different constant values?
* I don't see the point to store data per layout. One Vec per bind point should be enough ideally, which should also make pipeline inheritance easier (not handled so far as I can see)

Reply:
I love that kind of critique :)

> How does it handle the case where a user uses the same layout for compute and graphics, but pushes different constant values?

I don't know how this is needed to be handled, hence was asking you on gitter :) Are push constants associated with the (pipeline layout, bind point) pair? I thought it's just (pipeline layout), given that `vkCmdPushConstants` doesn't specify the bind point.

So in my understanding (and the current implementation) if the user pushes different constant values for compute, they will get them for graphics as well. Please tell me if it's supposed to work differently.

> I don't see the point to store data per layout. One Vec per bind point should be enough ideally, which should also make pipeline inheritance easier (not handled so far as I can see)

The idea was to handle the following use case:
  1. user binds a pipeline using layout A
  2. user pushes constants to layout B
  3. user binds a pipeline using layout B

If we only store data per bind point, we'd not be able to keep track of data associated with layouts not being bound.

> make pipeline inheritance easier (not handled so far as I can see)

Yeah, I haven't looked into inheritance really, yet. It should be fine to follow-up.

Reply:
>  if the user pushes different constant values for compute, they will get them for graphics as well. Please tell me if it's supposed to work differently.

Double checked the spec on this and this appears to be correct in case of push constants as the user needs to specify all shader flags, which means both will be set anyway at the same time.

> If we only store data per bind point, we'd not be able to keep track of data associated with layouts not being bound.

I think I should have explained it a bit more in detail. Up to my understand binding descriptor sets  works like this:
We have a small contiguous range of memory which stores push constant data or descriptor table pointers. This memory range is hidden from us and duplicated for each of the bind points (one graphics and one compute). On an action command (draw/dispatch) the shaders will use the layout of the current pipeline for interpreting and fetching the user data (constants and descriptor table pointers) from the associated memory range. When binding a descriptor set we need the layout that the driver can decide at which point in the memory range the tables should be written. Changing the layout during bind calls has no real influence beside that your data might be messed up/invalidates. Pipeline layout compatibility specifies which bindings are still valid. Therefore a pipeline layout should just be seen of an interpretation scheme (and abstraction layer).

In an example:
```
    user binds descriptors to layout A
    user binds descriptors to layout B
    user binds a pipeline using layout B
    draw
    user binds a pipeline using layout A
    draw
```
The 2nd call will overwrite (usually) the binding of layout A. Which means the users need to rebind the descriptor sets to layout A before the draw call due to invalidation. If A and B are compatible (up to the descriptor sets by A) then no rebinding is required.